test/e2e/node_reuse_test.go

```diff
@@ -0,0 +1,574 @@
+	Eventually(
+		func() int {
+			machines := &clusterv1.MachineList{}
+			Expect(targetClusterClient.List(ctx, machines, client.InNamespace(namespace))).To(Succeed())
+			running_upgraded_count := 0
+			for _, machine := range machines.Items {
+				if machine.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
+					running_upgraded_count++
+				}
+			}
+			return running_upgraded_count
+		},
+		e2eConfig.GetIntervals(specName, "wait-machine-running")...,
+	).Should(Equal(NUMBER_OF_ALL_BMH))
+}
+
+func scaleMachineDeployment(ctx context.Context, clusterClient client.Client, oldReplicas int, newReplicas int) {
+	machineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+		Lister:      clusterClient,
+		ClusterName: clusterName,
+		Namespace:   namespace,
+	})
+	Expect(len(machineDeployments)).To(Equal(1), "Expected exactly 1 MachineDeployment")
+	machineDeploy := machineDeployments[0]
+	Expect(*machineDeploy.Spec.Replicas).To(Equal(int32(oldReplicas)), "Expect the exact old replicas (%v) of MachineDeployment", oldReplicas)
+	patch := []byte(fmt.Sprintf(`{"spec": {"replicas": %d}}`, newReplicas))
+	err := clusterClient.Patch(ctx, machineDeploy, client.RawPatch(types.MergePatchType, patch))
+	Expect(err).To(BeNil(), "Failed to patch workers MachineDeployment")
+}
+
+func scaleControlPlane(ctx context.Context, c client.Client, name client.ObjectKey, newReplicaCount int) {
+	ctrlplane := kcp.KubeadmControlPlane{}
+	Expect(c.Get(ctx, name, &ctrlplane)).To(Succeed())
+	helper, err := patch.NewHelper(&ctrlplane, c)
+	Expect(err).To(BeNil(), "Failed to create new patch helper")
+
+	ctrlplane.Spec.Replicas = pointer.Int32Ptr(int32(newReplicaCount))
+	Expect(helper.Patch(ctx, &ctrlplane)).To(Succeed())
+}
+
+func untaintNodes(clientSet *kubernetes.Clientset, nodes *corev1.NodeList, taint *corev1.Taint) {
+	for i := range nodes.Items {
+		newNode, changed, err := taints.RemoveTaint(&nodes.Items[i], taint)
+		Expect(err).To(BeNil(), "Failed to remove taint")
+		if changed {
+			_, err = clientSet.CoreV1().Nodes().Update(ctx, newNode, metav1.UpdateOptions{})
+			Expect(err).To(BeNil(), "Failed to update nodes")
+		}
+	}
+}
+
+func annotateBmh(ctx context.Context, client client.Client, host bmo.BareMetalHost, key string, value *string) {
+	helper, err := patch.NewHelper(&host, client)
+	Expect(err).NotTo(HaveOccurred())
+	annotations := host.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	if value == nil {
+		delete(annotations, key)
+	} else {
+		annotations[key] = *value
+	}
+	host.SetAnnotations(annotations)
+	Expect(helper.Patch(ctx, &host)).To(Succeed())
+}
+
+func filterBmhsByProvisioningState(bmhs []bmo.BareMetalHost, state bmo.ProvisioningState) (result []bmo.BareMetalHost) {
+	for _, bmh := range bmhs {
+		if bmh.Status.Provisioning.State == state {
+			result = append(result, bmh)
+		}
+	}
+	return
+}
```

Comment:
IMO, other repetitive operations i.e updating m3mt node reuse field to true, maxSurgeField & k8s versions based if it is KCP or MD, UUID comparison after the upgrade could also be isolated and made a function as you did for others above